contracts/TokenHolder.sol

```diff
@@ -149,363 +171,347 @@ contract TokenHolder is MultiSigWallet {
-     * @param _nonce incremental nonce.
-     * @param _data the bytecode to be executed.
-     *         Use web3 getData method to construct _data.
-     * @param _v It's the recovery id.
-     * @param _r It's the output of an ECDSA signature.
-     * @param _s It's also the output of an ECDSA signature.
+     * @dev Function returns true if the key is authorized or was previously
+     *      authorized and revoked.
+     *      Function does not take into consideration the expiration height.
      *
-     * @return result is true/false.
+     * @return True in case if the key was authorized, otherwise false.
      */
-    function verifySignature(
-        address _from,
+    function wasEphemeralKeyAuthorized(address _ephemeralKey)
+        public
+        view
+        returns (bool)
+    {
+        return ephemeralKeys[_ephemeralKey].status !=
+            EphemeralKeyStatus.NON_AUTHORIZED;
+    }
+
+
+    /* Private Functions */
+
+    function processExecutableTransaction (
+        bytes4 _callPrefix,
         address _to,
-        uint256 _nonce,
         bytes _data,
+        uint256 _nonce,
         uint8 _v,
         bytes32 _r,
         bytes32 _s
     )
         private
-        returns (EphemeralKeyData storage ephemeralKeyData_)
+        returns (bytes32 messageHash_, address key_)
     {
-        require(
-            _from == address(this),
-            "From should be tokenholder contract address"
-        );
-        require(
-            _to != address(0),
-            "to address can't be 0"
-        );
-        require(
-            _nonce != 0,
-            "Nonce can't be 0."
-        );
-        require(
-            _data.length != 0,
-            "Data length can't be 0."
+        messageHash_ = getMessageHash (
+            _callPrefix,
+            _to,
+            keccak256(_data),
+            _nonce
         );
 
-        // Construct hashed message.
-        bytes32 messageHash = getHashedMessage(_from, _to, _data, _nonce);
-        address signer = ecrecover(messageHash, _v, _r, _s);
-        require(
-            isAuthorizedEphemeralKey(signer),
-            "Invalid ephemeral key!"
+        key_ = recoverKey (
+            messageHash_,
+            _v,
+            _r,
+            _s
         );
-        ephemeralKeyData_ = ephemeralKeys[signer];
-        require(
-            ephemeralKeyData_.expirationHeight >= block.number,
-            "ephemeral key has expired!"
+
+        require (
+            isEphemeralKeyActive(key_),
+            "Ephemeral key is not active."
         );
-        // Consume the nonce
-        ephemeralKeyData_.nonce = ephemeralKeyData_.nonce + 1;
-        require(
-            ephemeralKeyData_.nonce == _nonce,
-            "Invalid nonce!"
+
+        EphemeralKeyData storage keyData = ephemeralKeys[key_];
+
+        require (
+            _nonce == keyData.nonce,
+            "Nonce is not equal to the current nonce."
         );
```

Comment:
This check is not correct. User will pass next non consumed nonce from the interaction layer.
So first do 
keyData.nonce = keyData.nonce.add(1) and then
require (
            "Invalid nonce!"	            _nonce == keyData.nonce,
            "Nonce is not equal to the current nonce."
        );